staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -161,84 +165,35 @@ func (a *Allocator) Allocate(ctx context.Context, node *v1.Node) (finalResult []
-				return nil, fmt.Errorf("claim %s, request %s: could not retrieve device class %s: %w", klog.KObj(claim), request.Name, request.DeviceClassName, err)
+			if !a.prioritizedListEnabled && hasSubRequests {
+				return nil, fmt.Errorf("claim %s, request %s: has subrequests, but the feature is disabled", klog.KObj(claim), request.Name)
 			}
 
-			// Start collecting information about the request.
-			// The class must be set and stored before calling isSelectable.
-			requestData := requestData{
-				class: class,
-			}
-			requestKey := requestIndices{claimIndex: claimIndex, requestIndex: requestIndex}
-			alloc.requestData[requestKey] = requestData
-
-			switch request.AllocationMode {
-			case resourceapi.DeviceAllocationModeExactCount:
-				numDevices := request.Count
-				if numDevices > math.MaxInt {
-					// Allowed by API validation, but doesn't make sense.
-					return nil, fmt.Errorf("claim %s, request %s: exact count %d is too large", klog.KObj(claim), request.Name, numDevices)
+			var reqData requestData
+			if hasSubRequests {
+				reqData = requestData{
+					hasSubRequests: true,
 				}
-				requestData.numDevices = int(numDevices)
-			case resourceapi.DeviceAllocationModeAll:
-				requestData.allDevices = make([]deviceWithID, 0, resourceapi.AllocationResultsMaxSize)
-				for _, pool := range pools {
-					if pool.IsIncomplete {
-						return nil, fmt.Errorf("claim %s, request %s: asks for all devices, but resource pool %s is currently being updated", klog.KObj(claim), request.Name, pool.PoolID)
-					}
-					if pool.IsInvalid {
-						return nil, fmt.Errorf("claim %s, request %s: asks for all devices, but resource pool %s is currently invalid", klog.KObj(claim), request.Name, pool.PoolID)
-					}
-
-					for _, slice := range pool.Slices {
-						for deviceIndex := range slice.Spec.Devices {
-							selectable, err := alloc.isSelectable(requestKey, slice, deviceIndex)
-							if err != nil {
-								return nil, err
-							}
-							if selectable {
-								device := deviceWithID{
-									id:    DeviceID{Driver: slice.Spec.Driver, Pool: slice.Spec.Pool.Name, Device: slice.Spec.Devices[deviceIndex].Name},
-									basic: slice.Spec.Devices[deviceIndex].Basic,
-									slice: slice,
-								}
-								requestData.allDevices = append(requestData.allDevices, device)
-							}
-						}
-					}
+			} else {
+				reqData, err = alloc.validateDeviceRequest(requestAccessor{request: request}, requestKey, pools, hasSubRequests)
+				if err != nil {
+					return nil, err
 				}
-				requestData.numDevices = len(requestData.allDevices)
-				alloc.logger.V(6).Info("Request for 'all' devices", "claim", klog.KObj(claim), "request", request.Name, "numDevicesPerRequest", requestData.numDevices)
-			default:
-				return nil, fmt.Errorf("claim %s, request %s: unsupported count mode %s", klog.KObj(claim), request.Name, request.AllocationMode)
 			}
-			alloc.requestData[requestKey] = requestData
-			numDevicesPerClaim += requestData.numDevices
```

Comment:
After removing the code which increases `numDevicesPerClaim`, that variable is always zero and should have also been removed.

But I think it has value to keep it as a lower bound and thus will add it back in a PR against your branch.

This raises a question: does the code check the maximum number of allocation results now when evaluating alternatives? I haven't checked.

We need a test case for this:
- req 0:
  - alternative 0: resourceapi.AllocationResultsMaxSize devices
  - alternative 1: resourceapi.AllocationResultsMaxSize / 2 devices
- req 1:
  - alternative 0: resourceapi.AllocationResultsMaxSize devices
  - alternative 1: resourceapi.AllocationResultsMaxSize / 2 devices

There must 2 *  resourceapi.AllocationResultsMaxSize devices in the pool.

This has a solution (allocating the second alternative in both cases) and the allocator has to pick it.

Reply:
Found it:
```
	numDevicesAfterAlloc := len(alloc.result[r.claimIndex].devices) + requestData.numDevices
	if numDevicesAfterAlloc > resourceapi.AllocationResultsMaxSize {
		return false, fmt.Errorf("claim %s: number of requested devices %d exceeds the claim limit of %d", klog.KObj(claim), numDevicesAfterAlloc, resourceapi.AllocationResultsMaxSize)
	}
```

This cannot be a hard error anymore because that would abort the search for a solution.

Reply:
https://github.com/mortent/kubernetes/pull/1/ restores checking during validation with a lower bound estimate, but does not fix the issue described above.

Reply:
I added one test case for this, but I think we need another that covers the situation where the combination of requests might cause us to exceed the limit, but where there is a combination that will work. Will try to add that tomorrow.

Reply:
The scenario with a large number (`2 * resourceapi.AllocationResultsMaxSize`) of devices (that happen to be identical) where we run into the max number of allocated devices is actually pretty interesting. The current algorithm ends up trying a very large number of allocations that are all essentially equivalent, since replacing one device with another doesn't make a difference when the problem is the number of allocated devices.

I'm looking at ways to improve this. Currently thinking about a strategy where we can backtrack more aggressively when this happens. It might also be an option to do some kind of pre-flight checks before the actual allocation where we can identify combinations that will always exceed the limit and then avoid trying those during the allocation phase. But leaning towards the first approach for now.

Reply:
Added logic to avoid retrying different devices when we know that the next request or all subrequests failed with the allocation result size limit exceeded error. In this situation, trying different devices in the previous request will not make a difference, since the number of devices will not change. In situations with a large number of devices, the algorithm will end up trying a very large number of solutions that we know are infeasible.

Added the code and a test for this in a separate commit: https://github.com/kubernetes/kubernetes/pull/128586/commits/108385a08e12f2fab8a263598367973b2aaf3102

Reply:
Before `firstAlternativeOf`, the final number of devices was know in advance and was checked, so this particular issue did not cause unnecessary and ultimately failing attempts to find a solution.

But the algorithm was "dumb" in several ways, like for example when match attributes cause rejection of a solution: it would simply try a different device, even if that device has the same attributes as the one that just got rejected. I am not sure yet how to solve this in a way that doesn't lead to spaghetti code with lots of special cases.

I still need to look at your new commit - in the new year...